[REVIEW] Add in callbacks to give feedback on memory usage

revans2 opened the pull request:
This provides similar functionality to what dask does for tracking memory usage.

jlowe commented on java/src/main/java/ai/rapids/cudf/CSVOptions.java:
```diff
@@ -91,6 +97,18 @@ public static Builder builder() {
+     * @param guess the number of rows to be loaded as a guess.
+     * @return this for chaining.
+     */
+    public Builder withRowGuess(int guess) {
```
Does a row count make the most sense to use as a guess here?  Thinking of the string case and other variable-width types that will foil this approach.  Wondering if it should just be a memory size guess to let the caller have more control over the expected size if they happen to know more about the expected row contents than the schema would imply.  cudf could provide a utility method to return the minimum size of a schema (i.e.: just using the fixed-width types in the schema) and the user could control the fudging from there.

revans2 replied:
This is one thing I was thinking about and I am not totally sure what the right thing to do is. We can ask users to provide a size guess which is very generic, but it is partly pushing the burden back onto them.  Do they know the schema before it is loaded?  Possibly. If they do are they aware of approximately how much memory each row will take based off of that schema? You mention strings as an example but I don't think they know any better than us how large the string data really is. The end user may have some knowledge but we have other knowledge and to get a final imperfect guess we have to combine them together.  rows felt like something that the user might be able to come close to guessing at.

jlowe replied:
What I think would work best is having the user specify the memory size estimate.  In the case of the Spark plugin loading Parquet and ORC, the plugin knows far more about what's about to be loaded than the cudf layer does.  Therefore putting the guessing logic in cudf doesn't make sense to me.  The plugin knows the full schema and row count, so it can come up with a fairly close estimate if the schema is just fixed-width types.  This also allows it to control how variable-width types would be estimated (e.g.: controllable via a config settable by the end-user).

If cudf hardcodes the guessing logic then trying to control the memory size estimate on the caller's end becomes messy.  The caller would need to understand how cudf translates row counts to a memory size so the caller can fudge the reported row count accordingly to get the desired size.  If we keep the estimate as a row size then arguably there should be a way to specify the variable-width fields estimate so the caller has better control over the memory estimate when it knows more than cudf does.  At that point I think it's simpler to have only one API to have the caller specify the memory size directly.

> it is partly pushing the burden back onto them.

Yes, but cudf could provide a utility function to calculate a minimum row size for a schema. That would offload some of the calculation burden from the caller and keep the specifics of type sizes within cudf.  The caller would then be responsible for determining the sizes of variable-width fields to get the final average row size, and maybe it has access to some data that helps there.

jlowe commented on java/src/main/java/ai/rapids/cudf/Table.java:
```diff
@@ -656,8 +684,11 @@ public static Table readParquet(ParquetOptions opts, HostMemoryBuffer buffer,
   public static Table readORC(ORCOptions opts, File path) {
-    return new Table(gdfReadORC(opts.getIncludeColumnNames(),
-        path.getAbsolutePath(), 0, 0, opts.usingNumPyTypes(), opts.timeUnit().getNativeId()));
+    long amount = Math.max(path.length(), opts.getRowGuess() * 100);
```
Why `* 100` here?  I would expect the schema to be examined to compute the row-size estimate.  If the schema is all fixed-width types then we know how big it should be.  As it is now, it could be wildly off in either direction if the row count is high enough and the user can't easily compensate.  There's no way to override this via a config or other dynamic setting.

If we go with the total size estimate rather than row count estimate approach this isn't an issue.

revans2 replied:
The issue is that there is no schema passed into the ORC reader.  i went with row count because it was something again that an end user might reasonable be able to come up with. And like I said if you want me to change it I can, but I doubt it will produce a more accurate result in the long term.

revans2 replied:
After thinking it over I'll update the APIs to always take a size guess, that if provided we will use.  I'll also provide one, or possibly more APIs to help with the guess based off of row count and schema.

revans2 commented:
@jlowe I think I have addressed your current set of review comments

jlowe commented on java/src/main/java/ai/rapids/cudf/Table.java:
```diff
@@ -34,6 +34,10 @@
+  // This is an estimate of how compressed data is when guessing the output size of data
+  // For ORC and Parquet this is relatively conservative. We might want a different
+  // one for text based formats.
+  private static final long COMPRESSION_RATION_ESTIMATE = 10;
```
```suggestion
  private static final long COMPRESSION_RATIO_ESTIMATE = 10;
```

jlowe commented on java/src/main/java/ai/rapids/cudf/Table.java:
```diff
@@ -1000,14 +1069,17 @@ public Table aggregate(Aggregate... aggregates) {
+        aggregate = new Table(gdfGroupByAggregate(
+            operation.table.nativeHandle,
+            operation.indices,
+            // one way of converting List[Integer] to int[
```
```suggestion
            // one way of converting List[Integer] to int[]
```

jlowe commented on java/src/main/native/include/jni_utils.hpp:
```diff
@@ -28,6 +28,7 @@
 #include "cudf/legacy/table.hpp"
 #include "utilities/column_utils.hpp"
 #include "cudf/utilities/error.hpp"
+#include "utilities/legacy/error_utils.hpp"
```
Include only change in this file?

revans2 replied:
yes CUDA_TRY moved files and was breaking the build.

jlowe commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -706,6 +741,7 @@ void copyHostBufferBytes(byte[] dst, int dstOffset, BufferType src, long srcOffs
    * @return - Boolean vector
    */
   public ColumnVector isNotNull() {
+    // No prediction possible because built up from other operators.
```
I'm not following the comment here.  We know how many rows and thus approximately how big the resulting vector would be.  Is this trying to say the accounting is already handled within `validityAsBooleanVector` or...?

jlowe replied:
Still wondering if this is saying it is impossible to make a prediction here (not sure why that would be the case?) or if this is saying the prediction is already being handled by the method being called.

revans2 commented on java/src/main/java/ai/rapids/cudf/MemoryListener.java:
```diff
@@ -0,0 +1,137 @@
+ */
+public abstract class MemoryListener {
+  /**
+   * A prediction about how much memory is about to be used for an operation.
```
I'm not sure how to do it properly with bookkeeping.  Do we add the reservations together?  Do later reservations take away from the previous ones?  No nesting just makes things so much simpler.

revans2 commented on java/src/main/java/ai/rapids/cudf/MemoryListener.java:
```diff
@@ -0,0 +1,137 @@
+
+  /**
+   * Indicates that the previous prediction on this thread is no longer in force. Predictions should
+   * never nest.  Only one prediction should ever be in force at a time.
```
I thought it would be up to the end user to decide what they want to do with the information.  They can keep track of the exact memory that has been allocated and ignore the predictions, or they can use the predictions to make a reservation and then deduct memory from that reservation as it is allocated.  It all depends on what you want to do.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -306,7 +314,9 @@ public ColumnVector getLengths() {
    */
   public ColumnVector hash() {
-    return new ColumnVector(hash(getNativeCudfColumnAddress(), HashFunction.MURMUR3.nativeId));
+    try (DevicePrediction prediction = new DevicePrediction(DType.INT32.sizeInBytes * rows, "hash")) {
```
These are guesses and we can always work to make them more accurate. I am happy to add a validity vector to this one if we want to, but it is really going to ultimately end up being a game of whack a mole.

jlowe replied:
I ask because other parts of this patch accounted for the validity vector.  It would be nice to account for it if it's easy to do so, but OK to punt this to a followup if it's not straightforward.

revans2 commented:
@jlowe I have addressed all of your comments that I felt I could. If you think it is still lacking please let me know and I will try to expand on the comments more.

kuhushukla commented on java/src/main/java/ai/rapids/cudf/DevicePrediction.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Provides a convenient way to setup a prediction for a device data operation. The prediction
+ * should by part of a try with resources block so that close is guaranteed to be called and
+ * end the prediction. Do not nest these make sure that the prediction is as close to the actual
```
nit : _to_ make sure ?

revans2 commented:
@kuhushukla I have addressed your comment.

jlowe commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -743,6 +781,7 @@ public ColumnVector findAndReplaceAll(ColumnVector oldValues, ColumnVector newVa
    */
   public ColumnVector isNull() {
     ColumnVector res = null;
+    // No prediction is possible because it is built up from other operators
```
I see the comment in `isNotNull` was updated, I assume the same needs to be done here.

revans2 replied:
I thought your issue was that the comment was not obvious.  In this one isNull is built up from other operators.  `validityAsBooleanVector` and the `not`.  If it is still confusing I can expand the comment out.

jlowe replied:
My issue is with the `No prediction is possible` part -- it implies that this operation can never make a prediction.  How about something like:
```suggestion
    // Memory prediction handled by underlying operations
```

abellina commented on java/src/main/java/ai/rapids/cudf/Schema.java:
```diff
@@ -55,24 +57,61 @@ public static Builder builder() {
+        total += type.sizeInBytes * numRows;
+      }
+      // Assume that there is validity
+      total += (numRows + 7)/8;
```
nit, space around `/`

abellina commented on java/src/main/java/ai/rapids/cudf/Table.java:
```diff
@@ -638,8 +681,14 @@ public static Table readParquet(ParquetOptions opts, HostMemoryBuffer buffer,
     assert offset >= 0 && offset < buffer.length;
-    return new Table(gdfReadParquet(opts.getIncludeColumnNames(),
-        null, buffer.getAddress() + offset, len, opts.timeUnit().getNativeId()));
+    long amount = opts.getSizeGuess();
```
just thinking out loud, could `getSizeGuess` be `getSizeGuessOrElse(len * COMPRESSION_RATIO_ESTIMATE)` for example?

abellina commented on java/src/main/java/ai/rapids/cudf/Table.java:
```diff
@@ -748,13 +809,17 @@ public static Table concatenate(Table... tables) {
       assert tables[i].getNumberOfColumns() == numColumns : "all tables must have the same schema";
     }
-    return new Table(concatenate(tableHandles));
+    try (DevicePrediction prediction = new DevicePrediction(amount,"concat")) {
```
nit, space after `,`

abellina commented on java/src/main/java/ai/rapids/cudf/MemoryListener.java:
```diff
@@ -0,0 +1,147 @@
+   * operations.
+   * @param listener the listener to start sending events to.
+   */
+  public static void registerDeviceListener(final MemoryListener listener) {
```
should this be `registerMemoryListener`?

revans2 replied:
No because I wanted to potentially reuse the same interface in the future for pinned host memory or just regular host memory

revans2 commented:
@abellina I believe I have addressed all of your review comments please take another look.

revans2 commented:
@jlowe I think I have addressed your final comment with this last commit.
